File: Workspace/Feedback Control for Computer Systems/src/main/java/nl/tudelft/rvh/chapter1/BufferOpen.java
package nl.tudelft.rvh.chapter1;

import java.util.concurrent.TimeUnit;

import nl.tudelft.rvh.ChartTab;
import nl.tudelft.rvh.Tuple;
import rx.Observable;
import rx.functions.Func1;

public class BufferOpen extends ChartTab {

	public BufferOpen() {
		super("Chapter 1 - open", "Open simulation of random buffer simulation",
				"time", "queue size");
	}

	@Override
	public String seriesName() {
		return "closed simulation";
	}

	@Override
	public Observable<Tuple<Number, Number>> runSimulation() {
		Observable<Long> time = Observable.interval(1L, TimeUnit.MILLISECONDS).take(5000);
		Buffer b = new Buffer(10, 10);
		Func1<Long, Double> target = t -> 5.0;

		Observable<Integer> map = time.map(target).map(b::work);
		return time.zipWith(map, Tuple<Number, Number>::new);
	}
}

File: Workspace/Feedback Control for Computer Systems/src/main/scala/nl/tudelft/rvh/chapter1/BufferOpen.scala
package nl.tudelft.rvh.chapter1

import scala.concurrent.duration.DurationInt
import nl.tudelft.rvh.ChartTab
import rx.lang.scala.Observable
import rx.lang.scala.schedulers.ComputationScheduler

class BufferOpen extends ChartTab("Chapter 1 - open", "Open simulation of random buffer simulation", "time", "queue size") {

	def seriesName = "closed simulation"
	
	override def time = Observable interval (1 millisecond, ComputationScheduler()) take 5000
	
	def setpoint(time: Long) = 5.0

	def simulation = {
		val buffer = new Buffer(10, 10)
		time.map(setpoint).map(buffer.work(_))
	}
}